### PR TITLE
af-gen-event.[h|cpp] removal

### PR DESCRIPTION
**Pull request by rochaferraz**
Removes usage of `af-gen-event.h` across project. This touches

- build files, especially some CMake build files that had specific references to autogenerated `zzz_generated/<sample>` folder
- Control Cluster
- Door Lock Server
- Barrier Control Server
- All af-gen-event.h files removed

This refactoring is a spin-off from PR the https://github.com/project-chip/connectedhomeip/pull/24654

See issue https://github.com/project-chip/connectedhomeip/issues/16785

### CODE REVIEW

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 1]
```diff
@@ -23,11 +23,49 @@
...
+void ColorControlServer::timerCallback(System::Layer *, void * callbackContext)
+{
+    auto control = static_cast<EmberEventControl *>(callbackContext);
+    (control->callback)(control->endpoint);
+}
+
+void ColorControlServer::schedule(EmberEventControl * control, uint32_t delayMs)
+{
+    CHIP_ERROR err = DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(delayMs), timerCallback,
+                                                           reinterpret_cast<void *>(control));
```
Comment (Damian-Nordic) [inline src/app/clusters/color-control-server/color-control-server.cpp:46]: `reinterpret_cast` is needed. Cast to `void*` is implicit.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 2]
```diff
@@ -49,6 +52,42 @@ OnOffServer OnOffServer::instance;
...
 static OnOffEffect * inst(EndpointId endpoint);
 
+/**********************************************************
+ * Matter timer scheduling glue logic
+ *********************************************************/
+
+void OnOffServer::timerCallback(System::Layer *, void * callbackContext)
+{
+    auto control = static_cast<EmberEventControl *>(callbackContext);
+    (control->callback)(control->endpoint);
```
Comment (Damian-Nordic) [inline src/app/clusters/on-off-server/on-off-server.cpp:62]: Hmm. If all these functions still refer to `EmberEventControl`, would it make sense to implement them in a single place and reuse?

Comment (Damian-Nordic) [reply src/app/clusters/on-off-server/on-off-server.cpp:62]: Ah, I see. That's fine with me. I was just curious.

File: src/app/util/attribute-storage.cpp [hunk 3]
```diff
@@ -869,17 +869,9 @@ bool emberAfEndpointEnableDisable(EndpointId endpoint, bool enable)
...
-                //                           endpoint,
-                //                           cluster->clusterId,
-                //                           ((cluster->mask & CLUSTER_MASK_CLIENT)
-                //                            ? "client"
-                //                            : "server"));
-                //        emberAfCoreFlush();
-                emberAfDeactivateClusterTick(
-                    endpoint, cluster->clusterId,
-                    (cluster->mask & CLUSTER_MASK_CLIENT ? EMBER_AF_CLIENT_CLUSTER_TICK : EMBER_AF_SERVER_CLUSTER_TICK));
+                // Note: Matter has no central "tick" engine for clusters
```
Comment (andy31415) [inline src/app/util/attribute-storage.cpp:872]: should we comment out the entire loop?

File: src/app/clusters/barrier-control-server/barrier-control-server.cpp [hunk 4]
```diff
@@ -51,6 +54,28 @@ static State state;
...
+ *********************************************************/
+
+void emberAfBarrierControlClusterServerTickCallback(EndpointId endpoint);
+
+static void timerCallback(System::Layer *, void * callbackContext)
+{
+    emberAfBarrierControlClusterServerTickCallback(static_cast<EndpointId>(reinterpret_cast<uintptr_t>(callbackContext)));
+}
+
+static void schedule(EndpointId endpoint, uint32_t delayMs)
```
Comment (andy31415) [inline src/app/clusters/barrier-control-server/barrier-control-server.cpp:68]: I wonder about `schedule/deactivate` naming ... they don't seem obvious.

How about `scheduleTimerCallackMs` and `deactivateEndpointTimerCallback` ?

File: src/app/clusters/color-control-server/color-control-server.h [hunk 5]
```diff
@@ -190,6 +190,12 @@ class ColorControlServer
     void computePwmFromXy(chip::EndpointId endpoint);
     bool computeNewColor16uValue(Color16uTransitionState * p);
 
+    // Matter timer scheduling glue logic
+    static void timerCallback(chip::System::Layer *, void * callbackContext);
+    void schedule(EmberEventControl * control, uint32_t delayMs);
```
Comment (andy31415) [inline src/app/clusters/color-control-server/color-control-server.h:195]: similar comment here: maybe `scheduleTimerCallbackMs` and `deactivateTimerCallbackForEndpoint` / `deactivateTimerCallbackForControlPoint`  (or even `deactivateTimerCallback` with overloads ... although generally I find it odd to understand relation between endpointId and EmberEventControl)

File: src/app/clusters/barrier-control-server/barrier-control-server.cpp (hunk 4 above)
Comment (rochaferraz) [reply src/app/clusters/barrier-control-server/barrier-control-server.cpp]: renamed, ptal

File: src/app/clusters/barrier-control-server/barrier-control-server.cpp [hunk 6]
```diff
@@ -51,6 +54,28 @@ static State state;
...
+    emberAfBarrierControlClusterServerTickCallback(static_cast<EndpointId>(reinterpret_cast<uintptr_t>(callbackContext)));
+}
+
+static void scheduleTimerCallbackMs(EndpointId endpoint, uint32_t delayMs)
+{
+    DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(delayMs), timerCallback,
+                                          reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+static void deactivateEndpointTimerCallback(EndpointId endpoint)
```
Comment (bzbarsky-apple) [inline src/app/clusters/barrier-control-server/barrier-control-server.cpp:74]: I wonder whether this should be cancelEndpointTimerCallback, but either way.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 7]
```diff
@@ -23,11 +23,48 @@
...
+}
+
+void ColorControlServer::scheduleTimerCallbackMs(EmberEventControl * control, uint32_t delayMs)
+{
+    CHIP_ERROR err = DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(delayMs), timerCallback, control);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Color Control Server failed to schedule event: %" CHIP_ERROR_FORMAT, err.Format());
+    }
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:50]: On success, doesn't this function need to set `control->status = EMBER_EVENT_MS_TIME` to be compatible with what the old code did?

As a followup (separate PR) we can look into getting rid of the `status` member, I think.....

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 8]
```diff
@@ -3425,18 +3420,14 @@ void DoorLockServer::SendLockOperationEvent(chip::EndpointId endpointId, LockOpe
...
+    uint32_t timeoutMs = timeoutSec * MILLISECOND_TICKS_PER_SECOND;
+    CHIP_ERROR err =
+        DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(timeoutMs), DoorLockOnAutoRelockCallback,
+                                              reinterpret_cast<void *>(static_cast<uintptr_t>(endpointId)));
 
-    if (EMBER_SUCCESS != err)
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(Zcl, "Failed to schedule autorelock: timeout=%" PRIu32 ", status=0x%x", timeoutSec, err);
+        ChipLogError(Zcl, "Color Control Server failed to schedule event: %" CHIP_ERROR_FORMAT, err.Format());
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:3430]: Please fix the error message here.  This is not Color Control, and generally the message should match what used to be here.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 9]
```diff
@@ -49,6 +52,41 @@ OnOffServer OnOffServer::instance;
...
+}
+
+void OnOffServer::scheduleTimerCallbackMs(EmberEventControl * control, uint32_t delayMs)
+{
+    CHIP_ERROR err = DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(delayMs), timerCallback, control);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "OnOff Server failed to schedule event: %" CHIP_ERROR_FORMAT, err.Format());
+    }
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:72]: Again, on success should set the status of the control correctly.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 10]
```diff
@@ -147,12 +185,7 @@ EmberAfStatus OnOffServer::setOnOffValue(chip::EndpointId endpoint, chip::Comman
                 Attributes::OffWaitTime::Set(endpoint, 0);
 
                 // Stop timer on the endpoint
-                EmberEventControl * event = getEventControl(endpoint);
-                if (event != nullptr)
-                {
-                    emberEventControlSetInactive(event);
-                    emberAfOnOffClusterPrintln("On/Toggle Command - Stop Timer");
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:154]: Can we keep this logging instead of just losing it?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 11]
```diff
@@ -147,12 +185,7 @@ EmberAfStatus OnOffServer::setOnOffValue(chip::EndpointId endpoint, chip::Comman
                 Attributes::OffWaitTime::Set(endpoint, 0);
 
                 // Stop timer on the endpoint
-                EmberEventControl * event = getEventControl(endpoint);
-                if (event != nullptr)
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:151]: Why is this null-check OK to remove?

File: src/app/util/attribute-storage.cpp [hunk 12]
```diff
@@ -866,22 +866,6 @@ bool emberAfEndpointEnableDisable(EndpointId endpoint, bool enable)
...
-            {
-                const EmberAfCluster * cluster = &((emAfEndpoints[index].endpointType->cluster)[i]);
-                //        emberAfCorePrintln("Disabling cluster tick for ep:%d, cluster:0x%2X, %p",
-                //                           endpoint,
-                //                           cluster->clusterId,
-                //                           ((cluster->mask & CLUSTER_MASK_CLIENT)
-                //                            ? "client"
-                //                            : "server"));
-                //        emberAfCoreFlush();
-                emberAfDeactivateClusterTick(
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.cpp:880]: Don't we need some equivalent to cancel the timers for the relevant clusters for the relevant endpoint?

What is doing that now?  This needs to happen.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/util/attribute-storage.cpp (hunk 12 above)
Comment (rochaferraz) [reply src/app/util/attribute-storage.cpp:884]: created mechanism. Pls check https://github.com/project-chip/zap/pull/921

File: src/app/zap-templates/templates/app/att-storage.zapt [hunk 13]
```diff
@@ -13,11 +13,11 @@
 #define CLUSTER_MASK_DEFAULT_RESPONSE_FUNCTION (0x04)
 // Does this cluster have message sent function?
 #define CLUSTER_MASK_MESSAGE_SENT_FUNCTION (0x08)
-// Does this cluster have manufacturer specific attribute changed function?
-#define CLUSTER_MASK_MANUFACTURER_SPECIFIC_ATTRIBUTE_CHANGED_FUNCTION (0x10)
```
Comment (rochaferraz) [inline src/app/zap-templates/templates/app/att-storage.zapt:17]: This `CLUSTER_MASK_MANUFACTURER_SPECIFIC_ATTRIBUTE_CHANGED_FUNCTION` isn't used. Replacing it seems preferable to adding a new cluster mask, as we would need to change the mask to 16 bits

**Review by rochaferraz — COMMENTED** [review]
CI will only succeed after ZAP https://github.com/project-chip/zap/pull/921

File: src/app/clusters/level-control/level-control.cpp [hunk 14]
```diff
@@ -1287,6 +1292,11 @@ void emberAfLevelControlClusterServerInitCallback(EndpointId endpoint)
     emberAfPluginLevelControlClusterServerPostInitCallback(endpoint);
 }
 
+void MatterLevelControlClusterServerShutdownCallback(EndpointId endpoint)
+{
+    DeviceLayer::SystemLayer().CancelTimer(timerCallback, reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
```
Comment (bzbarsky-apple) [inline src/app/clusters/level-control/level-control.cpp:1297]: Why not call `cancelEndpointTimerCallback` here?

File: src/app/util/attribute-storage.cpp [hunk 15]
```diff
@@ -362,6 +363,57 @@ static void initializeEndpoint(EmberAfDefinedEndpoint * definedEndpoint)
...
+    uint8_t clusterIndex;
+    const EmberAfEndpointType * epType = definedEndpoint->endpointType;
+    for (clusterIndex = 0; clusterIndex < epType->clusterCount; clusterIndex++)
+    {
+        const EmberAfCluster * cluster = &(epType->cluster[clusterIndex]);
+        EmberAfGenericClusterFunction f;
+        f = emberAfFindClusterFunction(cluster, CLUSTER_MASK_SHUTDOWN_FUNCTION);
+        if (f != nullptr)
+        {
+            ((EmberAfInitFunction) f)(definedEndpoint->endpoint);
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.cpp:378]: This is a little confusing.  We should have a cast to `EmberAfShutdownFunction` here, and define that typedef in `src/app/util/af-types.h`.

File: src/app/zap-templates/templates/app/att-storage.zapt [hunk 16]
```diff
@@ -13,11 +13,11 @@
 #define CLUSTER_MASK_DEFAULT_RESPONSE_FUNCTION (0x04)
 // Does this cluster have message sent function?
 #define CLUSTER_MASK_MESSAGE_SENT_FUNCTION (0x08)
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/att-storage.zapt:15]: Please file a followup to get rid of all the unused bits here; we seem to have several.

**Review by bzbarsky-apple — APPROVED** [review]
Thank you for doing this!  It's nice to get rid of this long-standing pain point...

File: src/app/clusters/level-control/level-control.cpp (hunk 14 above)
Comment (rochaferraz) [reply src/app/clusters/level-control/level-control.cpp]: agree

File: src/app/util/attribute-storage.cpp (hunk 15 above)
Comment (rochaferraz) [reply src/app/util/attribute-storage.cpp]: done

File: src/app/zap-templates/templates/app/att-storage.zapt (hunk 16 above)
Comment (rochaferraz) [reply src/app/zap-templates/templates/app/att-storage.zapt:15]: done
https://github.com/project-chip/connectedhomeip/issues/24984

File: src/app/clusters/color-control-server/color-control-server.h [hunk 17]
```diff
@@ -240,7 +247,8 @@ class ColorControlServer
     Color16uTransitionState colorTempTransitionStates[EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT];
 #endif // EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
 
-    EmberEventControl eventControls[EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT];
+    EmberEventControl
+        eventControls[EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT];
```
Comment (rochaferraz) [inline src/app/clusters/color-control-server/color-control-server.h:251]: This seemed the right thing to do considering the dynamic eps, but happy to hear inputs

File: src/app/clusters/color-control-server/color-control-server.h [hunk 18]
```diff
@@ -240,7 +248,8 @@ class ColorControlServer
...
```
[the remark this answers is not in the document]
Comment (rochaferraz) [reply src/app/clusters/color-control-server/color-control-server.h:260]: ok

File: .github/workflows/examples-efr32.yaml [hunk 19]
```diff
@@ -88,25 +88,25 @@ jobs:
           ./scripts/run_in_build_env.sh \
              "./scripts/build/build_examples.py \
                 --enable-flashbundle \
-                --target efr32-brd4187c-window-covering \
-                --target efr32-brd4187c-switch \
-                --target efr32-brd4187c-switch-sed \
+                --target efr32-brd4187c-window-covering-additional_data_advertising \
+                --target efr32-brd4187c-switch-openthread_mtd-shell \
+                --target efr32-brd4187c-switch-sed-low-power-use_ot_coap_lib \
```
Comment (bzbarsky-apple) [inline .github/workflows/examples-efr32.yaml:93]: Why are there unrelated workflow changes here?

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 20]
```diff
@@ -3396,18 +3393,16 @@ void DoorLockServer::SendLockOperationEvent(chip::EndpointId endpointId, LockOpe
...
 
-    if (EMBER_SUCCESS != err)
+    CHIP_ERROR err =
+        DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(timeoutMs), DoorLockOnAutoRelockCallback,
+                                              reinterpret_cast<void *>(static_cast<uintptr_t>(endpointId)));
+
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(Zcl, "Failed to schedule autorelock: timeout=%" PRIu32 ", status=0x%x", timeoutSec, err);
+        ChipLogError(Zcl, "Failed to schedule autorelock: timeout=%" PRIu32 ", status=0x%" PRIx32, timeoutSec, err.AsInteger());
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:3405]:
```suggestion
        ChipLogError(Zcl, "Failed to schedule autorelock: timeout=%" PRIu32 ", status=%" CHIP_ERROR_FORMAT, timeoutSec, err.Format());
```

File: .github/workflows/examples-efr32.yaml (hunk 19 above)
Comment (rochaferraz) [reply .github/workflows/examples-efr32.yaml]: interesting how this crept in. Maybe during rebase. Fixed

Comment (bzbarsky-apple) [reply .github/workflows/examples-efr32.yaml]: Now I'm seeing unrelated changes in the opposite direction, removing changes that have landed....  This PR should just not be changing this file!

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 20 above)
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:3405]: This still has the AsInteger thing?  As in, the accepted suggestion got overridden by a force-push or something?

Comment (bzbarsky-apple) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:3405]: And note the bit about the workflow file above.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by arkq — APPROVED** [review]
